indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java

```diff
@@ -1946,53 +1956,152 @@ private boolean updatePartitionDataFromStream()
       }
     }
 
-    if (supportsPartitionExpiration()) {
-      // Look for expired shards and remove them from metadata storage and the partition groups
-      Set<PartitionIdType> expiredPartitions = new HashSet<>();
-      for (PartitionIdType partitionTd : closedPartitions) {
-        if (!partitionIds.contains(partitionTd)) {
-          expiredPartitions.add(partitionTd);
+    if (!partitionIds.equals(previousPartitionIds)) {
+      // the set of partition IDs has changed, have any running tasks stop early so that we can adjust to the
+      // repartitioning quickly by creating new tasks
+      for (TaskGroup taskGroup : activelyReadingTaskGroups.values()) {
+        if (!taskGroup.taskIds().isEmpty()) {
+          // a new partition was added and we are managing active tasks - set an early publish time 2 minutes in the
+          // future to give things time to settle
+          earlyStopTime = DateTimes.nowUtc().plusMinutes(2);
```

Comment:
Where did this number come from, should it be a configuration, or is there any thing else like this in other parts of the codebase?

Reply:
I updated this to use a configurable `repartitionTransitionDuration` property under the supervisor tuning config, along with docs on what the property is for (giving the stream time to write records to the new shards before transitioning to a new set of tasks)

Reply:
It makes sense to support that for Kafka as well (partitions can be added), but I'll do that in a follow-on PR